osu.Game/Database/UserLookupCache.cs

```diff
@@ -15,103 +14,73 @@ namespace osu.Game.Database
-            }
-
-            List<User> nextLookup()
-            {
-                int[] lookupItems;
-
-                lock (taskAssignmentLock)
-                {
-                    pendingRequestConsumedIDs = true;
-                    lookupItems = nextTaskIDs.ToArray();
-                    nextTaskIDs.Clear();
-
-                    if (lookupItems.Length == 0)
-                    {
-                        queueNextTask(null);
-                        return new List<User>();
-                    }
-                }
+                var tcs = new TaskCompletionSource<User>();
 
-                var request = new GetUsersRequest(lookupItems);
+                // Add to the queue.
+                pendingUserTasks.Enqueue((userId, tcs));
 
-                // rather than queueing, we maintain our own single-threaded request stream.
-                api.Perform(request);
+                // Create a request task if there's not already one.
+                if (pendingRequestTask == null)
+                    createNewTask();
 
-                return request.Result?.Users;
+                return tcs.Task;
             }
         }
 
-        /// <summary>
-        /// Queues new work at the end of the current work tasks.
-        /// Ensures the provided work is eventually run.
-        /// </summary>
-        /// <param name="work">The work to run. Can be null to signify the end of available work.</param>
-        /// <returns>The task tracking this work.</returns>
-        private Task<List<User>> queueNextTask(Func<List<User>> work)
+        private void performLookup()
         {
+            var userTasks = new List<(int id, TaskCompletionSource<User> task)>();
+
+            // Grab at most 50 users from the queue.
             lock (taskAssignmentLock)
             {
-                if (work == null)
-                {
-                    pendingRequest = null;
-                    pendingRequestConsumedIDs = false;
-                }
-                else if (pendingRequest == null)
-                {
-                    // special case for the first request ever.
-                    pendingRequest = Task.Run(work);
-                    pendingRequestConsumedIDs = false;
-                }
-                else
+                while (pendingUserTasks.Count > 0 && userTasks.Count < 50)
                 {
-                    // append the new request on to the last to be executed.
-                    pendingRequest = pendingRequest.ContinueWith(_ => work());
-                    pendingRequestConsumedIDs = false;
+                    (int id, TaskCompletionSource<User> task) next = pendingUserTasks.Dequeue();
+
+                    // Perform a secondary check for existence, in case the user was queried in a previous batch.
+                    if (CheckExists(next.id, out var existing))
+                        next.task.SetResult(existing);
+                    else
+                        userTasks.Add(next);
```

Comment:
Not 100% on the necessity of the `CheckExisting` call you have, but there's also of requesting the same user in the lookup twice (which was handled on my PR).

To fix this, we could either:

- Store the IDs in a separate `HashSet` alongside `userTasks` (ensures always sending a request for 50 users).
- call Distinct in the `userTasks.Select(t => t.id)` below (simpler to implement, but requests may be short of 50. probably doesn't matter)

leaning towards the second option.

Reply:
At the point where we're calling `Distinct`, we may as well just store a temporary hashset along with the main list and keep the up-to-50 users. Fine with that?

Reply:
ok